### DISCUSSION TITLE
What functions are included in the arpack-ng library?

### DISCUSSION THREAD

**Discussion by Jiangsese** [opencollab/arpack-ng#454] [Q&A]
I'm a beginner, and the Fortran program provided by my teacher uses the ARPACK and LAPACK libraries. I have installed the ARPACK library and called it along with the MKL library, but I encountered the following errors:

zndrv1.obj : error LNK2019: unresolved external symbol ZNAUPD, referenced in function ZNAUPD.t871p.t872p.t873p.t874p.t875p.t876p.t877p.t878p.t879p.t880p.t881p.t882p.t883p.t884p.t885p.t886p.t887p.t3v.t3v

zndrv1.obj : error LNK2019: unresolved external symbol ZNEUPD, referenced in function ZNEUPD.t965p.t966p.t967p.t968p.t969p.t970p.t971p.t972p.t973p.t974p.t975p.t976p.t977p.t978p.t979p.t980p.t981p.t982p.t983p.t984p.t985p.t986p.t987p.t988p.t3v.t3v.t3v

eigenhh1.obj : error LNK2019: unresolved external symbol DSAUPD, referenced in function DSAUPD.t544p.t545p.t546p.t547p.t548p.t549p.t550p.t551p.t552p.t553p.t554p.t555p.t556p.t557p.t558p.t559p.t3v.t3v

eigenhh1.obj : error LNK2019: unresolved external symbol DSEUPD, referenced in function DSEUPD.t633p.t634p.t635p.t636p.t637p.t638p.t639p.t640p.t641p.t642p.t643p.t644p.t645p.t646p.t647p.t648p.t649p.t650p.t651p.t652p.t653p.t654p.t3v.t3v.t3v

eigenhh1.obj : error LNK2019: unresolved external symbol DMOUT, referenced in function DNRM2.void

Could you please confirm if these functions are not included in the ARPACK-NG library, or if there might be an issue with my installation? Thank you for your response.

**Comment by Jiangsese** [comment]
Additionally, even if I don't use MKL or don't link the ARPACK-NG library, the same errors occur.

**Comment by fghoussen** [comment]
> Could you please confirm if these functions are not included in the ARPACK-NG library

All `[SDCZ]S[AE]UPD` methods are provided by arpack.
How they are designed is explained here: https://github.com/opencollab/arpack-ng/blob/4caa8f4c703c6408b05b878f376e5df762d9a0e9/README.md?plain=1#L365
MKL is tricky to use: start simple with regular BLAS/LAPACK first.

**Reply by Jiangsese** [reply]
However, I tried again. If only ARPACK-NG is called, the same errors still occur:

ssh.obj: error LNK2019: unresolved external symbol ZGEEV
eigenhh1.obj: error LNK2019: unresolved external symbol DSAUPD
eigenhh1.obj: error LNK2019: unresolved external symbol DSEUPD
eigenhh1.obj: error LNK2019: unresolved external symbol DAXPY
eigenhh1.obj: error LNK2019: unresolved external symbol DNRM2
eigenhh1.obj: error LNK2019: unresolved external symbol DMOUT
zndrv1.obj: error LNK2019: unresolved external symbol ZNAUPD
zndrv1.obj: error LNK2019: unresolved external symbol ZNEUPD
zndrv1.obj: error LNK2019: unresolved external symbol ZAXPY
zndrv1.obj: error LNK2019: unresolved external symbol DZNRM2
zndrv1.obj: error LNK2019: unresolved external symbol DLAPY2

Is this due to not correctly including the ARPACK library, or is it a normal occurrence because LAPACK is not called? If I simultaneously call MKL's LAPACK and this ARPACK-NG library, only 5 errors occur:

eigenhh1.obj: error LNK2019: unresolved external symbol DSAUPD
eigenhh1.obj: error LNK2019: unresolved external symbol DSEUPD
eigenhh1.obj: error LNK2019: unresolved external symbol DMOUT
zndrv1.obj: error LNK2019: unresolved external symbol ZNAUPD
zndrv1.obj: error LNK2019: unresolved external symbol ZNEUPD

I'm sorry, I don't know how to format the response properly. This is my first time using GitHub.

**Reply by Jiangsese** [reply]
I use the Visual studio 2022 in Windows11

**Comment by fghoussen** [comment]
> unresolved external symbol 

You must link with the library: `gfortran ... -larpack`

**Comment by fghoussen** [comment]
>  Visual studio 2022 in Windows11

Not sure arpack can be built on Windows the windows way (with IDE like VS) but it should be possible the unix way (using msys) https://github.com/opencollab/arpack-ng/blob/4caa8f4c703c6408b05b878f376e5df762d9a0e9/.github/workflows/jobs.yml#L253-L293

**Reply by Jiangsese** [reply]
I'll try it latter and thank you very much for your help.

**Comment by fghoussen** [comment]
> Not sure arpack can be built on Windows the windows way (with IDE like VS)

...Maybe using `vcpkg`? @FabienPean @juanjosegarciaripoll Can you help @Jiangsese on this? (no idea how to do this as I'am debian based). If so, do not hesitate to detail the doc: https://github.com/opencollab/arpack-ng?tab=readme-ov-file#windows-support

**Reply by FabienPean** [reply]
Microsoft does not provide Fortran compiler but Intel does, which I believe integrate inside VS, so arpack could be built there. The install via vcpkg relies (at least then) on MSYS2/MinGW64. The issue is not detailed but as you noted above it is most likely a linker issue.

**Reply by fghoussen** [reply]
@FabienPean : do not hesitate to detail the doc: https://github.com/opencollab/arpack-ng?tab=readme-ov-file#windows-support (how to install compilers on windows? is msys2/mingw64 mandatory? ...)
